pysradb/search.py

```diff
@@ -0,0 +1,437 @@
+        MissingQueryException.
+
+    Methods
+    -------
+    get_df()
+        Returns the dataframe storing this search result
+
+    """
+
+    def __init__(self, verbosity, return_max, fields):
+        self.verbosity = verbosity
+        self.return_max = return_max
+        allowed_fields = ["query", "accession", "organism", "layout", "mbases", "publication_date", "platform",
+                          "selection", "source", "strategy", "title"]
+        for field in allowed_fields:
+            if field not in fields:
+                fields[field] = None
+        for k in fields:
+            if type(fields[k]) == list:
+                fields[k] = " ".join(fields[k])
+        if not any(fields):
+            raise MissingQueryException()
+        self.fields = fields
+        self.df = pd.DataFrame()
+
+    def search(self):
+        pass
+
+    def get_df(self):
+        return self.df
+
+
+class SraSearch(QuerySearch):
+    """Subclass of QuerySearch that implements search by querying
+    NCBI Entrez API
+
+    Methods
+    -------
+    search()
+        sends the user query via requests to NCBI Entrez API and returns
+        search results as a pandas dataframe.
+
+    _format_query_string()
+        formats the input user query into a string
+
+    _format_request()
+        formats the request payload
+
+    _format_result(content)
+        formats the search query output.
+
+    See Also
+    --------
+    QuerySearch: Superclass of SraSearch
+
+    """
+
+    def search(self):
+        # Step 1: retrieves the list of uids that satisfies the input
+        # search query
+
+        payload = self._format_request()
+        try:
+            r = requests_3_retries().get(
+                "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi",
+                params=OrderedDict(payload),
+                timeout=20,
+            )
+            r.raise_for_status()
+            uids = r.json()["esearchresult"]["idlist"]
+
+            # Step 2: retrieves the detailed information for each uid returned, in groups of 500.
+            if not uids:
+                print(
+                    f"No results found for the following search query: \n {self.fields}"
+                )
+                return  # If no queries found, return nothing
+
+            for i in range(0, len(uids), 300):
+                current_uids = ",".join(uids[i : min(i + 300, len(uids))])
+
+                payload2 = {"db": "sra", "retmode": "xml", "id": current_uids}
+
+                r = requests_3_retries().get(
+                    "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi",
+                    params=OrderedDict(payload2),
+                    timeout=20,
+                )
+                r.raise_for_status()
+                self._format_result(r.content)
+        except requests.exceptions.Timeout:
+            print(f"Connection to the server has timed out. Please retry.")
```

Comment:
This is a design issue that I guess we should discuss (ping @saketkc ), but personally I think Timeout and HTTPError exceptions should not be caught. Say you put this in a script, and the script finishes with exit code 0, then you might think: "Alright, no results for this query", when actually the query failed.

Reply:
I agree with @mvdbeek. This is currently failing silently. A script using this will also fail silently.